From 8835659a3fe2d2cf36f4dc2d09e5e68ad01d5b99 Mon Sep 17 00:00:00 2001
From: Anatoliy Bilenko <anatoliy_bilenko@xyratex.com>
Date: Thu, 14 Mar 2013 11:34:09 +0200
Subject: [PATCH 22/63] WARNING: This commit is used to reproduce the following
 issue:

 - NOTICE : [rpc/slot.c:584:m0_rpc_slot_reply_received] < rc=-71
 - Use the following command to reporduce:
   M0_TRACE_IMMEDIATE_MASK=rpc ./utils/ub.sh -t rpc-ub
---
 m0t1fs/linux_kernel/st/st |  2 +-
 rpc/ub/ub.c               | 46 ++++++++++++++--------------------------------
 2 files changed, 15 insertions(+), 33 deletions(-)

diff --git a/m0t1fs/linux_kernel/st/st b/m0t1fs/linux_kernel/st/st
index aeceefb..d5de11a 100755
--- a/m0t1fs/linux_kernel/st/st
+++ b/m0t1fs/linux_kernel/st/st
@@ -16,7 +16,7 @@ die() { echo "$@" >&2; exit 1; }
 
 M0_CORE_DIR=`readlink -f $0`
 M0_CORE_DIR=${M0_CORE_DIR%/*/*/*/*}
-# [[ "${M0_CORE_DIR##*/}" =~ ^mero ]] || die 'Unable to determine M0_CORE_DIR'
+[[ "${M0_CORE_DIR##*/}" =~ ^mero ]] || die 'Unable to determine M0_CORE_DIR'
 
 ## Path to the file with configuration string for confd.
 CONF_FILE=$SANDBOX_DIR/conf.txt
diff --git a/rpc/ub/ub.c b/rpc/ub/ub.c
index d89c499..92b382e 100644
--- a/rpc/ub/ub.c
+++ b/rpc/ub/ub.c
@@ -36,13 +36,6 @@
  * Client and Server definitions
  * ---------------------------------------------------------------- */
 
-#if 0
-#define CLIENT_ENDPOINT_ADDR    "0@lo:12345:34:*"
-#define SERVER_ENDPOINT_ADDR    "0@lo:12345:34:1"
-#define SERVER_ENDPOINT         "lnet:" SERVER_ENDPOINT_ADDR
-static struct m0_net_xprt *g_xprt = &m0_net_lnet_xprt;
-#endif
-
 enum {
 	CLIENT_COB_DOM_ID  = 16,
 	SESSION_SLOTS      = 15,
@@ -58,32 +51,18 @@ enum {
    - define rpc client interface such that it can represent multiple clients.
  */
 
+#if 0
+#define CLIENT_ENDPOINT_ADDR    "0@lo:12345:34:*"
+#define SERVER_ENDPOINT_ADDR    "0@lo:12345:34:1"
+#define SERVER_ENDPOINT         "lnet:" SERVER_ENDPOINT_ADDR
+
+static struct m0_net_xprt *g_xprt = &m0_net_lnet_xprt;
+#else
 #define CLIENT_ENDPOINT_ADDR "127.0.0.1:12345"
 #define SERVER_ENDPOINT_ADDR CLIENT_ENDPOINT_ADDR
 #define SERVER_ENDPOINT      "bulk-mem:" SERVER_ENDPOINT_ADDR
 
 static struct m0_net_xprt  *g_xprt = &m0_net_bulk_mem_xprt;
-
-#define NAME(ext) "rpc-ub" ext
-#if 0
-static struct m0_net_domain client_net_dom;
-static struct m0_dbenv      client_dbenv;
-static struct m0_cob_domain client_cob_dom;
-
-static struct m0_rpc_client_ctx g_cctx = {
-	.rcx_net_dom               = &client_net_dom,
-	.rcx_cob_dom               = &client_cob_dom,
-	.rcx_cob_dom_id            = CLIENT_COB_DOM_ID,
-	.rcx_dbenv                 = &client_dbenv,
-	.rcx_db_name               = NAME(".client-db"),
-	.rcx_nr_slots              = SESSION_SLOTS,
-	/* @todo: the following fields can be used as a prototype: */
-	.rcx_local_addr            = CLIENT_ENDPOINT_ADDR,
-	.rcx_remote_addr           = SERVER_ENDPOINT_ADDR,
-	.rcx_timeout_s             = CONNECT_TIMEOUT,
-	.rcx_max_rpcs_in_flight    = MAX_RPCS_IN_FLIGHT,
-	.rcx_recv_queue_min_length = M0_NET_TM_RECV_QUEUE_DEF_LEN
-};
 #endif
 
 /**
@@ -98,6 +77,7 @@ struct ub_rpc_client {
 
 static struct ub_rpc_client g_client[1];
 
+#define NAME(ext) "rpc-ub" ext
 static char *g_argv[] = {
 	NAME(""), "-r", "-p", "-T", "AD", "-D", NAME(".db"),
 	"-S", NAME(".stob"), "-A", NAME(".addb-stob"), "-e", SERVER_ENDPOINT,
@@ -177,8 +157,10 @@ static void _start(void)
 	M0_ASSERT(rc == 0);
 
 	for (i = 0; i < ARRAY_SIZE(g_client); ++i) {
-		rc = _client_init(&g_client[i], i + CLIENT_COB_DOM_ID,
-				  ".client-db", SESSION_SLOTS);
+		char clientdb[20];
+		snprintf(clientdb, ARRAY_SIZE(clientdb), ".client-db-%d", i);
+		rc = _client_init(&g_client[i], CLIENT_COB_DOM_ID + i,
+				  clientdb, SESSION_SLOTS);
 		M0_ASSERT(rc == 0);
 	}
 
@@ -262,11 +244,11 @@ static void XXX_name_me(int iter M0_UNUSED)
 	int rc;
 	struct m0_rpc_session *session;
 
-	/* @todo: For some reason the followng error may occur here:
+	/* @todo: For some reason the following error may occur here:
 	   mero: NOTICE : [rpc/slot.c:584:m0_rpc_slot_reply_received] < rc=-71.
 	   Needs investigation!
 	 */
-	for (i = 0; i < 3; ++i) {
+	for (i = 0; i < 30; ++i) {
 		for (j = 0; j < ARRAY_SIZE(g_client); ++j) {
 			session = &g_client[j].rc_ctx.rcx_session;
 			fop_send(session, i);
-- 
1.8.3.2

